app/models/webhook.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+class Webhook < ApplicationRecord
+  include Authorizable
+
+  EVENT_POSTFIX = ".#{Foreman::Observable::DEFAULT_NAMESPACE}"
+
+  EVENT_ALLOWLIST = %w[
+    host_created host_updated host_destroyed
+    hostgroup_created hostgroup_updated hostgroup_destroyed
+    user_created user_updated user_destroyed
+    domain_created domain_updated domain_destroyed
+    subnet_created subnet_updated subnet_destroyed
+    build_entered build_exited status_changed
+  ].map { |e| e + EVENT_POSTFIX }.freeze
+
+  DEFAULT_PAYLOAD_TEMPLATE = 'Webhook Payload - Payload Default'
+
+  attribute :events, :string, array: true, default: []
+
+  validates_lengths_from_database
+  validates :name, :target_url, :events, presence: true
+  validates :target_url, format: URI.regexp(%w[http https])
+
+  belongs_to :payload_template, foreign_key: :payload_template_id
+
+  before_save :set_default_template, unless: :payload_template
+
+  scope :for_event, ->(events) { where('events @> ARRAY[?]::varchar[]', Array(events)) }
+
+  def self.available_events
+    ::Foreman::EventSubscribers.all_observable_events & EVENT_ALLOWLIST
+  end
+
+  def self.deliver(event_name:, payload:)
+    for_event(event_name).each do |target|
+      target.deliver(event_name: event_name, payload: payload)
+    end
+  end
```

Comment:
The `@>` is costly operation, tablescan on *every* event which occurs in Foreman. I know that for small tables tablescan is usually forced even if there was a join table, but this smells slow. I wish there was a way to add a payload to the subscriber (event id) so every single webhook could subscribe and ID would easily find it. But after looking into Rails codebase it does not appear so.

I think we can probably keep this array approach but if and only if we use Rails.cache for each event fired storing list of webhook ids. Expiration time could be 10 minutes with explicit expiration when a webhook is created, updated or deleted.

Reply:
Before creating an optimilization ticket, let's solve the big question. Do we want to have ability to subscribe to multiple events for a single webhook? I only see one particular usecase - to create some kind of one-catch-them-all example for debugging, but overall business-wise I don't see use cases. @timogoebel what's your take on this? Can we refactor the array to a simple string?

Reply:
Just not to forget: https://projects.theforeman.org/issues/30796